loyalty_outlier_analysis.ipynb

```diff
@@ -0,0 +1,476 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "aggregators = {\n",
+    "    'purchase_amount': ['sum', 'mean', 'min', 'max', 'std', 'count'],\n",
+    "    'installments': ['sum', 'mean', 'min', 'max', 'std'],\n",
+    "    'month_lag': ['mean', 'min', 'max'],\n",
+    "    'merchant_id': ['nunique'],\n",
+    "    'merchant_category_id': ['nunique'],\n",
+    "    'state_id': ['nunique'],\n",
+    "    'city_id': ['nunique'],\n",
+    "    'subsector_id': ['nunique'],\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "add_aggregated_numerical_fields(train_and_validation_df, hist_trans_df, aggregators=aggregators)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "add_aggregated_categorical_fields(train_and_validation_df,\n",
+    "                                  hist_trans_df,\n",
+    "                                  column_names=['authorized_flag', 'category_1', 'category_2', 'category_3'])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# category_2 and category_3 contain nan values, so let's skip those for now.\n",
+    "add_top_categories(train_and_validation_df,\n",
+    "                   hist_trans_df,\n",
+    "                   column_names=['authorized_flag', 'category_1', 'subsector_id', 'city_id', 'state_id'])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Outlier Separation\n",
+    "\n",
+    "Outliers are the ones where the target value is below -30. Most target values are clustered around 0, but then we have that bunch of values around -33, with the same value. And apparently predicting those correctly is crucial for good performance in the competition."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "train_and_validation_df['outlier'] = np.where(train_and_validation_df['target'] < -30, 'Outlier', 'Normal')\n",
+    "train_and_validation_df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Separate the outliers into their own table"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "outlier_df = train_and_validation_df.loc[train_and_validation_df['outlier'] == 'Outlier'].copy()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "outlier_df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Trying to see if there are any differences in the correlations between the full dataset and the outlier dataset. Practically everything looks exactly the same."
```

Comment:
Another way to look at it would be like so:

```py
plt.figure(figsize = (14, 10))
sns.heatmap(train_and_validation_df.corr() - outlier_df.corr(), vmin=-1, vmax=1, cmap='PiYG')
```

Giving:

<img width="950" alt="screenshot 2019-01-29 at 17 36 04" src="https://user-images.githubusercontent.com/1210335/51923964-68dbcc80-23ec-11e9-86dc-a22a8a449b02.png">

The more interesting differences seem to be with `month_lag_mean` and `month_lag_min`.

Reply:
That's very interesting! Should I add it to this PR?

Reply:
If you want – as you prefer!